src/c/keeper.c

```diff
@@ -0,0 +1,796 @@
+#include "keeper.h"
+#include "service.h"
+#include "errorlist.h"
+#include "bus.h"
+#include "devsdk/devsdk.h"
+#include "iot/time.h"
+#include "api.h"
+#include "iot/base64.h"
+
+
+/* Our "impl" structure for global state. */
+typedef struct keeper_impl_t 
+{
+    devsdk_service_t *service;
+    iot_threadpool_t *pool;
+    iot_logger_t *lc;
+    char *host;
+    uint16_t port;
+    char *key_root;
+    char *topic_root;
+    devsdk_registry_updatefn updater;
+    void *updatectx;
+} keeper_impl_t;
+
+static int32_t edgex_keeper_client_notify(void *impl, const iot_data_t *request, const iot_data_t *pathparams, const iot_data_t *params, iot_data_t **reply);
+#define CONF_PREFIX "edgex/v3/"
+
+void *devsdk_registry_keeper_alloc(devsdk_service_t *service)
+{
+    keeper_impl_t *rv;
+    rv = calloc(sizeof(keeper_impl_t), 1);
+    rv->service = service;
+    return rv;
+}
+
+static void edgex_keeper_client_free (void *impl)
+{
+  keeper_impl_t *keeper = (keeper_impl_t *)impl;
+  if (keeper)
+  {
+    if (keeper->host)
+    {
+        free (keeper->host);
+    }
+    if (keeper->key_root)
+    {
+        free(keeper->key_root);
+    }
+    if (keeper->topic_root)
+    {
+        free(keeper->topic_root);
+    }
+    if (impl)
+    {
+        free (impl);
+    }
+  }
+}
+
+static void *delayed_message_bus_connect(void *impl)
+{
+    keeper_impl_t *keeper = (keeper_impl_t *)impl;
+
+    iot_log_debug(keeper->lc, "Message bus wait thread starting");
+    while (true)
+    {
+        if (keeper->service && keeper->service->msgbus)
+        {
+            char *tree = malloc(strlen(keeper->topic_root) + 3);
+            strcpy(tree, keeper->topic_root);
+            if (tree[strlen(tree) - 1] == '/')
+            {
+                tree[strlen(tree) - 1] = '\0';
+            }
+            strcat(tree, "/#");
```

Comment:
MQTT wildcards are not functional in the current sdk. https://github.com/edgexfoundry/device-sdk-c/blob/b4a60c126f741d84c14ee104c3263f3235d33930/src/c/bus.c#L45-L54 However, this only affects the "listen for config changes" feature and does not break the core functionality of the device service. A separate github issue (https://github.com/edgexfoundry/device-sdk-c/issues/523) has been opened to address this problem.

Reply:
fixed in #523